dev/test/query.ts

```diff
@@ -773,6 +788,163 @@ describe('query interface', () => {
+    let attempts = 0;
+
+    // count of stream initializations that are successful
+    // and make progress (a document is received before error)
+    const initializationsWithProgress = 10;
+
+    // Receiving these error codes on a stream after the stream has received document data
+    // should result in the stream retrying indefinitely.
+    const retryableErrorCodes = [
+      Status.DEADLINE_EXCEEDED,
+      Status.UNAVAILABLE,
+      Status.INTERNAL,
+      Status.UNAVAILABLE,
+    ];
+
+    const overrides: ApiOverride = {
+      runQuery: x => {
+        ++attempts;
+
+        // Validate that runQuery is called with cursor of the lastReceivedDocument
+        // for all attempts except but the first.
+        if (attempts > 1) {
+          const docPath =
+            x?.structuredQuery?.startAt?.values?.[0].referenceValue || '';
+          const docId = docPath.substring(docPath.lastIndexOf('/'));
+          expect(docId).to.equal(
+            `/id-${Math.min(initializationsWithProgress, attempts - 1)}`
+          );
+          expect(x?.structuredQuery?.orderBy?.length).to.equal(1);
+          expect(x?.structuredQuery?.orderBy?.[0].field?.fieldPath).to.equal(
+            '__name__'
+          );
+        }
+
+        const streamElements = [];
+
+        streamElements.push(heartbeat(1000));
+
+        // For the first X initializations, the stream will see progress.
+        // For the X+1 attempt, the stream will not see progress before
+        // the error, so we expect the stream to close.
+        if (attempts <= initializationsWithProgress) {
+          streamElements.push(result(`id-${attempts}`));
+        }
+
+        // Create an error with one of the retryable error codes
+        const googleError = new GoogleError();
+        googleError.code =
+          retryableErrorCodes[attempts % retryableErrorCodes.length];
+        googleError.message = 'test error message';
+        streamElements.push(googleError);
+
+        return stream(...streamElements);
+      },
+    };
+
+    return createInstance(overrides).then(firestoreInstance => {
+      firestore = firestoreInstance;
+      return firestore
+        .collection('collectionId')
+        .get()
+        .then(() => {
+          throw new Error('Unexpected success in Promise');
+        })
+        .catch(err => {
+          expect(err.message).to.equal('test error message');
+
+          // 20 stream initialization with data that progresses the steam
+          // followed by an error. The final X+1 initialization does not get data
+          // before the error, so we may also get retries during stream initialization.
+          const initilizationRetries = withHeartbeat ? 1 : 5;
+          expect(attempts).to.equal(
+            initializationsWithProgress + initilizationRetries
+          );
+        });
+    });
+  }
+
+  it('handles retryable exception until progress stops with heartbeat', () => {
+    handlesRetryableExceptionUntilProgressStops(true);
```

Comment:
You need to `return` since this is async.
```suggestion
    return handlesRetryableExceptionUntilProgressStops(true);
```